.github/workflows/wasm.yml

```diff
@@ -0,0 +1,123 @@
+name: WebAssembly 
+
+on:
+  push:
+    branches: [ main, ci-sandbox ]
+  pull_request:
+    branches: [ '**' ]
+
+env:
+  ccache_basedir: ${{ github.workspace }}
+  ccache_dir: "${{ github.workspace }}/.ccache"
+  ccache_compilercheck: mtime
+  ccache_compress: 'true'
+  ccache_compresslevel: 9
+  ccache_maxsize: 200M
+  ccache_cmake: -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -DCMAKE_C_COMPILER_LAUNCHER=ccache
+  emsdk_version: 2.0.9
+
+jobs:
+  build-wasm:
+    name: "emscripten"
+    runs-on: ubuntu-latest
+    steps:
+
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          submodules: recursive
+
+      - name: Set ccache environment for emcc
+        run: |
+          echo "CCACHE_COMPILER_CHECK=${{ env.ccache_compilercheck }}" >> $GITHUB_ENV
+          echo "CCACHE_BASEDIR=${{ env.ccache_basedir }}" >> $GITHUB_ENV
+          echo "CCACHE_COMPRESS=${{ env.ccache_compress }}" >> $GITHUB_ENV
+          echo "CCACHE_COMPRESSLEVEL=${{ env.ccache_compresslevel }}" >> $GITHUB_ENV
+          echo "CCACHE_DIR=${{ env.ccache_dir }}" >> $GITHUB_ENV
+          echo "CCACHE_MAXSIZE=${{ env.ccache_maxsize }}" >> $GITHUB_ENV
+          # https://emscripten.org/docs/compiling/Building-Projects.html#using-a-compiler-wrapper
+          echo "EM_COMPILER_WRAPPER=ccache" >> $GITHUB_ENV
+          
+      # This need to be run before setup, so ccache build caching doesn't complain.
+      - name: Obtain emsdk sources
+        run: |
+            git clone --depth 1 https://github.com/emscripten-core/emsdk.git
+
+      - name: Cache-op for build-cache through ccache
+        uses: actions/cache@v2
+        with:
+          path: |
+              ${{ env.ccache_dir }}
+              ${{ github.workspace }}/emsdk/ccache/git-emscripten_64bit/
+          key: ccache-${{ github.job }}-${{ env.emsdk_version }}-${{ steps.ccache_vars.outputs.hash }}-${{ github.ref }}-${{ steps.ccache_vars.outputs.timestamp }}
+          restore-keys: |-
+            ccache-${{ github.job }}-${{ env.emsdk_version }}-${{ steps.ccache_vars.outputs.hash }}-${{ github.ref }}
+            ccache-${{ github.job }}-${{ env.emsdk_version }}-${{ steps.ccache_vars.outputs.hash }}
+            ccache-${{ github.job }}-${{ env.emsdk_version }}
+
+      - name: Setup Emscripten toolchain
+        run: |
+            (cd emsdk && ./emsdk install ${{ env.emsdk_version }} ccache-git-emscripten-64bit)
+            (cd emsdk && ./emsdk activate ${{ env.emsdk_version }} ccache-git-emscripten-64bit)
+            # mtime of this file is checked by ccache, we set it to avoid cache misses.
+            touch -m -d '1 Jan 2021 12:00' emsdk/.emscripten
+
+            # These needs to be done in the activated shell.
+            eval $(./emsdk/emsdk construct_env \
+                | sed 's/export PATH=\(.*\);/echo \1 >> $GITHUB_PATH;/' \
+                | sed 's/export \(.*\);/echo \1 >> $GITHUB_ENV;/' );
```

Comment:
Why does this need to be wrapped in `eval $(…)`?

Reply:
```bash
#!/bin/bash
# script.sh

GITHUB_ENV=github.env
GITHUB_PATH=github.path

rm $GITHUB_ENV $GITHUB_PATH

./emsdk/emsdk activate ccache-git-emscripten-64bit 2.0.9 &> /dev/null || echo "Failed activation"

$1 $(./emsdk/emsdk construct_env  \
  | sed 's/export PATH=\(.*\);/echo \1 >> $GITHUB_PATH;/' \
  | sed 's/export \(.*\);/echo \1 >> $GITHUB_ENV;/')
```

```bash
$ bash script.sh 
tail: cannot open 'github.env' for reading: No such file or directory
tail: cannot open 'github.path' for reading: No such file or directory

$ bash script.sh eval
==> github.env <==
EMSDK=/home/jerin/versioned/sandbox/emsdk
EM_CONFIG=/home/jerin/versioned/sandbox/emsdk/.emscripten
EMSDK_NODE=/home/jerin/versioned/sandbox/emsdk/node/14.18.2_64bit/bin/node

==> github.path <==
/home/jerin/versioned/sandbox/emsdk:/home/jerin/versioned/sandbox/emsdk/upstream/emscripten:/home/jerin/versioned/sandbox/emsdk/node/14.18.2_64bit/bin:/home/jerin/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/usr/lib/wsl/lib:/mnt/c/Program Files/Alacritty/:/mnt/c/WINDOWS/system32:/mnt/c/WINDOWS:/mnt/c/WINDOWS/System32/Wbem:/mnt/c/WINDOWS/System32/WindowsPowerShell/v1.0/:/mnt/c/WINDOWS/System32/OpenSSH/:/mnt/c/Program Files/CMake/bin:/mnt/c/Users/jerin/AppData/Local/Microsoft/WindowsApps:/mnt/c/Users/jerin/AppData/Local/Programs/Microsoft VS Code/bin:/mnt/c/Program Files/Git/cmd:/snap/bin
```